chanfitness/revenue.go

```diff
@@ -0,0 +1,227 @@
+package chanfitness
+
+import (
+	"errors"
+
+	"github.com/btcsuite/btcd/wire"
+	"github.com/lightningnetwork/lnd/channeldb"
+	"github.com/lightningnetwork/lnd/lnwire"
+)
+
+var (
+	// maxQueryEvents is the number of events to query the forward log for at
+	// a time.
+	maxQueryEvents uint32 = 500
+
+	// errUnknownChannelID is returned when we cannot map a short channel ID to
+	// a channel outpoint.
+	errUnknownChannelID = errors.New("cannot find channel outpoint")
+)
+
+// eventsQuery is a function which returns paginated queries for forwarding
+// events.
+type eventsQuery func(
+	offset, maxEvents uint32) (channeldb.ForwardingLogTimeSlice, error)
+
+// GetRevenueReport calculates accumulates relevant forwarding logs and
+// produces a revenue report.
+func GetRevenueReport(channelIDs map[lnwire.ShortChannelID]wire.OutPoint,
+	query eventsQuery, attributeIncoming float64) (*RevenueReport, error) {
+
+	events, err := getEvents(channelIDs, query)
+	if err != nil {
+		return nil, err
+	}
+
+	return revenueReport(events, attributeIncoming), nil
+}
+
+// getEvents gets calls the paginated query function until it has all the
+// forwarding events for the period provided. It takes a map of shortChannelIDs
+// to outpoints which is used to convert forwarding events short ids to
+// outpoints.
+func getEvents(channelIDs map[lnwire.ShortChannelID]wire.OutPoint,
+	query eventsQuery) ([]revenueEvent, error) {
+	var (
+		offset uint32
+		events []revenueEvent
+	)
+
+	for {
+		forwarding, err := query(offset, maxQueryEvents)
+		if err != nil {
+			return nil, err
+		}
+
+		// Convert the event's short channel id to an outpoint and add it to
+		// our events list.
+		for _, fwd := range forwarding.ForwardingEvents {
+			incoming, ok := channelIDs[fwd.IncomingChanID]
+			if !ok {
+				return nil, errUnknownChannelID
+			}
+
+			outgoing, ok := channelIDs[fwd.OutgoingChanID]
+			if !ok {
+				return nil, errUnknownChannelID
```

Comment:
seems like this is likely to happen? If there are events for channels we are currently not querying for.

Reply:
The map should have all of our channels in it. It's not very efficient, but we don't have a way of determining which channels we have open over a timestamped period, so I thought it would be ok like this?